library/pod-security-policy/seccomp/template.yaml

```diff
@@ -42,45 +42,89 @@ spec:
     - target: admission.k8s.gatekeeper.sh
       rego: |
         package k8spspseccomp
-
+        
         import data.lib.exempt_container.is_exempt
-
-        violation[{"msg": msg, "details": {}}] {
-            metadata := input.review.object.metadata
-            not input_wildcard_allowed(metadata)
-            container := input_containers[_]
+        
+        container_annotation_key_prefix = "container.seccomp.security.alpha.kubernetes.io/"
+        
+        pod_annotation_key = "seccomp.security.alpha.kubernetes.io/pod"
+        
+        violation[{"msg": msg}] {
+            not input_wildcard_allowed
+            container := input_containers[name]
             not is_exempt(container)
-            not input_container_allowed(metadata, container)
-            msg := sprintf("Seccomp profile is not allowed, pod: %v, container: %v, Allowed profiles: %v", [metadata.name, container.name, input.parameters.allowedProfiles])
+            result := get_profile(container)
+            not allowed_profile(result.profile)
+            msg := sprintf("Seccomp profile '%v' is not allowed for container '%v'. Found at: %v. Allowed profiles: %v", [result.profile, name, result.location, input.parameters.allowedProfiles])
         }
-
-        input_wildcard_allowed(metadata) {
+        
+        input_wildcard_allowed {
             input.parameters.allowedProfiles[_] == "*"
         }
-
-        input_container_allowed(metadata, container) {
-            not get_container_profile(metadata, container)
-            metadata.annotations["seccomp.security.alpha.kubernetes.io/pod"] == input.parameters.allowedProfiles[_]
+        
+        allowed_profile(profile) {
+            profile == input.parameters.allowedProfiles[_]
         }
-
-        input_container_allowed(metadata, container) {
-          profile := get_container_profile(metadata, container)
-          profile == input.parameters.allowedProfiles[_]
+        
+        # Container profile as defined in pod annotation
+        get_profile(container) = {"profile": profile, "location": location} {
+            not has_annotation(get_container_annotation_key(container.name))
```

Comment:
What happens if both the annotation and the security context are defined?

Reply:
I did some testing of these scenarios. Thankfully the `kube-apiserver` has protections for the case where the securityContext and annotation don't match by sending this error:
```
The Pod "nginx-seccomp-not-match" is invalid: spec.containers[0].securityContext.seccompProfile.type: Forbidden: seccomp type in annotation and field must match
```
Then I confirmed that the preference for container specific over pod default works even across the two options (i.e. container annotation is used over the pod default in securityContext). So I have added protections to the get_profile functions to ensure only one returns in all cases. I have added a preference to the securityContext version over the annotation only because annotations are deprecated but as stated before they must match before getting to the admission controller.

I've added unit test coverage for these.